File: clients/kotlin/src/main/kotlin/com/oresoftware/flags2env/Flags2Env.kt
package com.oresoftware.flags2env.kotlin

import com.oresoftware.flags2env.Flags2Env as JavaFlags2Env

object Flags2Env {
    @JvmStatic
    fun parse(argv: List<String>, configPath: String? = null): Map<String, String> {
        val items = argv.map { it.toString() }.toTypedArray()
        return if (configPath == null) JavaFlags2Env.parse(items) else JavaFlags2Env.parse(configPath, items)
    }

    @JvmStatic
    fun parseProcess(configPath: String? = null): Map<String, String> {
        return if (configPath == null) JavaFlags2Env.parseProcess() else JavaFlags2Env.parseProcess(configPath)
    }

    @JvmStatic
    fun apply(env: Map<String, String>, argv: List<String>, configPath: String? = null): Map<String, String> {
        return env + parse(argv, configPath)
    }

    @JvmStatic
    fun applyProcess(env: Map<String, String>, configPath: String? = null): Map<String, String> {
        return env + parseProcess(configPath)
    }
}

File: clients/scala/src/main/scala/com/oresoftware/flags2env/Flags2Env.scala
package com.oresoftware.flags2env.scala

import com.oresoftware.flags2env.{Flags2Env => JavaFlags2Env}
import scala.jdk.CollectionConverters._

object Flags2Env {
  def parse(argv: Seq[String], configPath: Option[String] = None): Map[String, String] = {
    val items = argv.map(_.toString).toArray
    val parsed = configPath match {
      case Some(path) => JavaFlags2Env.parse(path, items)
      case None => JavaFlags2Env.parse(items)
    }
    parsed.asScala.toMap
  }

  def parseProcess(configPath: Option[String] = None): Map[String, String] = {
    val parsed = configPath match {
      case Some(path) => JavaFlags2Env.parseProcess(path)
      case None => JavaFlags2Env.parseProcess()
    }
    parsed.asScala.toMap
  }

  def apply(env: Map[String, String], argv: Seq[String], configPath: Option[String] = None): Map[String, String] =
    env ++ parse(argv, configPath)

  def applyProcess(env: Map[String, String], configPath: Option[String] = None): Map[String, String] =
    env ++ parseProcess(configPath)
}
